pkg/kv/kvserver/kvflowcontrol/replica_rac2/admission.go

```diff
@@ -0,0 +1,277 @@
+			interval{first: first, last: last, lowPriOverride: lowPriOverride})
+	}
+	if newLeader {
+		// Drop all intervals starting at or after the first index. Do it from the
+		// back, so that the append-only case is the fast path. When a suffix of
+		// entries is overwritten, the cost of this loop is an amortized O(1).
+		for ; n != 0 && p.intervals[n-1].first >= first; n-- {
+		}
+		p.intervals = p.intervals[:n]
+		// INVARIANT: keep > 0 => p.intervals[keep-1].first < first.
+
+		// Adjust the last interval if it overlaps with the one being added.
+		if n != 0 && p.intervals[n-1].last >= first {
+			// Truncate the last interval.
+			p.intervals[n-1].last = first - 1
+		}
+		extendOrAppend()
+		return
+	}
+	// Common case: existing leader.
+	if n > 0 {
+		if p.intervals[n-1].last >= last {
+			return
+		}
+		// INVARIANT: p.intervals[n-1].last < last, so p.intervals[n-1].last + 1 <= last.
+		if p.intervals[n-1].last >= first {
+			first = p.intervals[n-1].last + 1
+		}
+	}
+	extendOrAppend()
+}
+
+func (p *lowPriOverrideState) getEffectivePriority(
+	index uint64, pri raftpb.Priority,
+) (effectivePri raftpb.Priority) {
+	// Garbage collect intervals ending before the given index.
+	drop := 0
+	for n := len(p.intervals); drop < n && p.intervals[drop].last < index; drop++ {
+	}
+	p.intervals = p.intervals[drop:]
+	n := len(p.intervals)
+	// INVARIANT: if n > 0, p.intervals[0].last >= index.
+
+	// If there is no interval containing the index, return the original
+	// priority. We need to tolerate this since this may be case C1, and this is
+	// index 9, so we ignored when sideChannelForLowPriOverride provided
+	// information for 9.
+	if n == 0 || p.intervals[0].first > index {
+		return pri
+	}
+	lowPriOverride := p.intervals[0].lowPriOverride
+	// Remove the prefix of indices <= index.
+	if p.intervals[0].last > index {
+		p.intervals[0].first = index + 1
+	} else {
+		p.intervals = p.intervals[1:]
+	}
+	if lowPriOverride {
+		return raftpb.LowPri
+	}
+	return pri
+}
+
+// waitingForAdmissionState records the indices of individual entries that are
+// waiting for admission in the AC queues. These are added after emerging as
+// MsgStorageAppend in handleRaftReady, hence we can expect monotonicity of
+// indices within a leader term, and if the term changes, we can assume that
+// non-monotonicity means raft log indices being overwritten. These
+// assumptions simplify the processing of add, since it can wipe out higher
+// indices.
+type waitingForAdmissionState struct {
+	// The indices for each priority are in increasing index order.
+	//
+	// Say the indices for a priority are 3, 6, 10. We track the individual
+	// indices since when 3 is popped, we can advance admitted for that priority
+	// to 5. When 6 is popped, admitted can advance to 9. We should never have a
+	// situation where indices are popped out of order, but we tolerate that by
+	// popping the prefix upto the index being popped.
+	waiting [raftpb.NumPriorities][]admissionEntry
```

Comment:
Seemingly, tracking `[]admissionEntry` is unnecessary, and we can track just a tuple of `{leaderTerm, last, admitted}`, where `last` is the last entry with the given priority, and `admitted` is the last admitted index at this priority.

That would mean we can't jump to admitted 5 in the example above (it would move to 3), but is it necessary? There are no entries between index 3 and 6, so admitted index advance to 5 doesn't release new things.

But we will be able to detect that all entries at the given priority are admitted, if `admitted == last`. Needed in the `computeAdmitted` method, instead of `if len(w.waiting[i]) > 0`.